### PR TITLE
Add contact segment membership webhook

### PR DESCRIPTION
**Pull request by kuzmany**
| Q                                      | A
| -------------------------------------- | ---
| Bug fix? (use the a.b branch)          | no
| New feature/enhancement? (use the a.x branch)      | yes
| Deprecations?                          | no
| BC breaks? (use the c.x branch)        | no
| Automated tests included?              | not yet
| Related user documentation PR URL      | 
| Related developer documentation PR URL | https://github.com/mautic/developer-documentation/pull/204
| Issue(s) addressed                     | 

#### Description:

This PR just add Contact Segment Membership Change Event to Webhooks

image

#### Steps to test this PR:

1. Load up [this PR](https://mautibox.com)
2. Go to https://webhook.site/ and copy link of your webhook
3. Go to Mautic and Webhooks
4. Create  Contact Segment Membership Change Event to your webhook.site url
5. Go to contact list and change segment membership for few contacts
6. Check If you receive correct data to webhook.site

<a href="https://gitpod.io/#https://github.com/mautic/mautic/pull/10511"></a>

### CODE REVIEW

**Review by escopecz — CHANGES_REQUESTED** [review]
The changes look good. The only thing I'm missing is a functional test that would ensure this works and continue working.

**Comment by kuzmany** [comment]
Unit test added

**Review by RCheesley — DISMISSED** [review]
Tested with both campaign based and manual segment changes and the webhooks were correctly fired with both. 

Thanks for the PR @kuzmany just a PHPSTAN fix needed it looks like! 🚀

**Comment by kuzmany** [comment]
@RCheesley done

**Comment by kuzmany** [comment]
@RCheesley done

File: app/bundles/LeadBundle/Tests/EventListener/WebhookSubscriberTest.php [hunk 1]
```diff
@@ -185,4 +188,28 @@ public function testOnCompanySaveAndDelete()
...
+
+        $webhookSubscriber = new WebhookSubscriber($mockModel);
+
+        $this->dispatcher->addSubscriber($webhookSubscriber);
+
+        $lead    = new Lead();
+        $segment = new LeadList();
+        $event   = new ListChangeEvent($lead, $segment, true);
+        $this->dispatcher->dispatch(LeadEvents::LEAD_LIST_CHANGE, $event);
+    }
```
Comment (escopecz) [inline app/bundles/LeadBundle/Tests/EventListener/WebhookSubscriberTest.php:214]: @kuzmany this test does not make much sense. You dispatch an event directly in the test and then assert that you dispatched the event. The only thing it tests is Symfony's dispatcher.

Could you instead write a functional test where you set the webhook config to send in background create a contact, a segment and a webhook, add the contact to the segment via the `ListModel::addLead()` method and then check that the payload is stored in the webhook queue?

**Review by escopecz — CHANGES_REQUESTED** [review]

File: app/bundles/LeadBundle/Tests/EventListener/WebhookSubscriberTest.php (hunk 1 above)
Comment (kuzmany) [reply app/bundles/LeadBundle/Tests/EventListener/WebhookSubscriberTest.php:214]: What do you mean check payload in functional tests?

Comment (escopecz) [reply app/bundles/LeadBundle/Tests/EventListener/WebhookSubscriberTest.php:214]: When the webhooks are configured to send by a background job then the payload isn't sent directly but stored in the `]webhook_queue` table instead waiting for a cron job to pick it up. So we can assert that the correct record in that table exists when the segment membership is changed.

Comment (RCheesley) [reply app/bundles/LeadBundle/Tests/EventListener/WebhookSubscriberTest.php:214]: @kuzmany are you able to make the suggested changes to the test?

Comment (kuzmany) [reply app/bundles/LeadBundle/Tests/EventListener/WebhookSubscriberTest.php:214]: Yes, I am able to do it.

Comment (kuzmany) [reply app/bundles/LeadBundle/Tests/EventListener/WebhookSubscriberTest.php:214]: I will check it later.

**Comment by RCheesley** [comment]
@cla-bot check

**Comment by RCheesley** [comment]
@kuzmany will need to be rebased to 5.x please, when you're ready!

**Comment by kuzmany** [comment]
Close favor to https://github.com/mautic/mautic/pull/12186

**Pull request closed without merging**